source/CustomSale.cpp

```diff
@@ -0,0 +1,407 @@
+/* CustomSale.cpp
+Copyright (c) 2021 by Hurleveur
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "CustomSale.h"
+
+#include "ConditionsStore.h"
+#include "DataNode.h"
+#include "GameData.h"
+#include "Logger.h"
+#include "Outfit.h"
+#include "Planet.h"
+#include "Set.h"
+
+using namespace std;
+
+namespace
+{
+	const auto show = map<CustomSale::SellType, const string> {
+		{CustomSale::SellType::DEFAULT, ""},
+		{CustomSale::SellType::IMPORT, "import"}
+	};
+	// Default value for finding prices and offsets, that should not conflict with actual values.
+	const double DEFAULT = numeric_limits<double>::infinity();
+}
+
+
+
+void CustomSale::Load(const DataNode &node, bool eventChange)
+{
+	const Set<Sale<Outfit>> &items = GameData::Outfitters();
+	const Set<Outfit> &outfits = GameData::Outfits();
+
+	bool isAdd = false;
+	const Outfit *outfit = nullptr;
+	// Outfitters or outfits mode.
+	const string mode = node.Token(1);
+	name = node.Token(2);
+	auto parseValueOrOffset = [&isAdd, &outfit, &mode, this](double &amount, const DataNode &line)
+	{
+		int size = line.Size();
+		// Default is 1, because outfits can be added only to have a custom sellType.
+		if(isAdd)
+			amount += size > 2 ? line.Value(2) : 1.;
+		else
+			amount = size > 1 ? line.Value(1) : 1.;
+		// All values are converted into pourcentages if that is not how they are given (which would be indicated by %)
+		// This means that the offset is handled as relative to the modified price instead of the default one.
+		// Outfitter changes always are pourcentages.
+		if((mode != "outfitters"
+				&& (size == (2 + isAdd)
+				|| (size > 2 && line.Token(2 + isAdd) != "%"))))
+			toConvert.push_back(make_pair(outfit, &amount));
+	};
+
+	for(const DataNode &child : node)
+	{
+		bool remove = child.Token(0) == "remove";
+		bool add = child.Token(0) == "add";
+
+		int keyIndex = (add || remove);
+		bool hasKey = child.Size() > 1;
+
+		const string &key = child.Token(keyIndex);
```

Comment:
`keyIndex` can be `1`, but you have not guaranteed that this node actually has a token at that index, and this will throw an exception if (it just does `std::vector::operator[index]`) if you don't bounds check (trust me, I know - #8017).